openssl/src/x509/mod.rs

```diff
@@ -443,6 +460,153 @@ impl X509Ref {
         }
     }
 
+    /// Returns this certificate's key usage.
+    ///
+    /// This corresponds to [`X509_get_ext_d2i`] called with `NID_key_usage`.
+    ///
+    /// [`X509_get_ext_d2i`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_d2i.html
+    pub fn key_usage(&self) -> Option<KeyUsageFlags> {
+        unsafe {
+            let key_usage_ptr = ffi::X509_get_ext_d2i(
+                self.as_ptr(),
+                ffi::NID_key_usage,
+                ptr::null_mut(),
+                ptr::null_mut(),
+            );
+
+            if key_usage_ptr.is_null() {
+                return None;
+            }
+
+            let key_usage_str = Asn1BitString::from_ptr(key_usage_ptr as *mut _);
+            if key_usage_str.len() == 0 {
+                return None;
+            }
+
+            let mut flags: u32 = 0;
+            let num_bytes = cmp::min(key_usage_str.len(), mem::size_of::<u32>());
+            let data = key_usage_str.as_slice();
+
+            for i in 0..num_bytes {
+                flags ^= (data[i] as u32) << ((num_bytes - (i + 1)) * 8);
```

Comment:
This can be done more easily with `u32::from_le_bytes` I think.

Reply:
In this same function I did change the return type to `KeyUsageFlags` since the OpenSSL API [`X509_get_key_usage`](https://www.openssl.org/docs/man1.1.0/man3/X509_get_key_usage.html) (not available in earlier versions) specifies: "If key usage is absent then UINT32_MAX is returned."

Since bitflags cannot contain bits not explicitly defined I've opted to just  return all of the flags we know about.